grove/tomography/utils.py

```diff
@@ -0,0 +1,640 @@
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+
+"""
+Utilities for encapsulating bases and properties of quantum operators and super-operators
+as represented by qutip.Qobj()'s.
+"""
+
+import itertools
+import logging
+from collections import OrderedDict, namedtuple
+from itertools import product as cartesian_product
+
+import matplotlib.pyplot as plt
+import matplotlib as mpl
+from matplotlib.colors import LinearSegmentedColormap
+import numpy as np
+import qutip as qt
+from mpl_toolkits.mplot3d import Axes3D
+from pyquil.quil import Program
+from pyquil.gates import I, X
+from scipy.sparse import hstack as sphstack, vstack as spvstack
+from scipy.sparse.linalg import norm as spnorm
+try:
+    # Python 2
+    from itertools import izip
+except ImportError:
+    # Python 3
+    izip = zip
+
+
+_log = logging.getLogger(__name__)
+
+THREE_COLOR_MAP = ['#48737F', '#FFFFFF', '#D6619E']
+rigetti_3_color_cm = LinearSegmentedColormap.from_list("Rigetti", THREE_COLOR_MAP[::-1], N=100)
+
+FIVE_COLOR_MAP = ['#C671A2', '#545253', '#85B5BE', '#ECE9CC', '#C671A2']
+rigetti_4_color_cm = LinearSegmentedColormap.from_list("Rigetti", FIVE_COLOR_MAP[::-1], N=100)
+
+FROBENIUS = 'fro'
+CHOI = 'choi'
+
+
+QX = qt.sigmax()
+QY = qt.sigmay()
+QZ = qt.sigmaz()
+QI = qt.qeye(2)
+
+EPS = 1e-8
+
+
+def sample_outcomes(probs, n):
+    """
+    For a discrete probability distribution ``probs`` with outcomes 0, 1, ..., k-1 draw ``n``
+    random samples.
+
+    :param list probs: A list of probabilities.
+    :param Number n: The number of random samples to draw.
+    :return: An array of samples drawn from distribution probs over 0, ..., len(probs) - 1
+    :rtype: numpy.ndarray
+    """
+    dist = np.cumsum(probs)
+    rs = np.random.rand(n)
+    return np.array([(np.where(r < dist)[0][0]) for r in rs])
+
+
+def basis_state_preps(*qubits):
+    """
+    Generate a sequence of programs that prepares the measurement
+    basis states of some set of qubits in the order such that the qubit
+    with highest index is iterated over the most quickly:
+    E.g., for qubits=(0, 1), it returns the circuits::
+
+        I_0 I_1
+        I_0 X_1
+        X_0 I_1
+        X_0 X_1
+    :param list qubits: Each qubit to include in the basis state preparation.
+    :return: Yields programs for each basis state preparation.
+    :rtype: Program
+    """
+    for prep in cartesian_product([I, X], repeat=len(qubits)):
+        basis_prep = Program()
+        for gate, qubit in zip(prep, qubits):
+            basis_prep.inst(gate(qubit))
+        yield basis_prep
+
+
+def basis_labels(n):
```

Comment:
consider using numpy functionality:
```
num = 100
str_len = 8
np.binary_repr(num, str_len)
```

Reply:
Ok, I've considered it, but I like this functional style of doing it. Also saves us precious microseconds :P
```
In [10]: def basis_labels_np(q):
    ...:     return [np.binary_repr(n, q) for n in range(2**q)]
    ...:

In [11]: %timeit basis_labels_np(4)
10000 loops, best of 3: 26.9 µs per loop

In [12]: %timeit basis_labels(4)
100000 loops, best of 3: 5.7 µs per loop
```

Reply:
sounds good.